lib/Builtins.cpp

```diff
@@ -14,171 +14,335 @@
+      return GetParameterType(mangled_name, type_info, pos);
+    } else if (type_code == 'h') { // OCL half
+      type_info->type_id = Type::FloatTyID;
+      type_info->is_signed = true;
+      type_info->byte_len = 2;
+    } else {
+#ifdef DEBUG
+      printf("Func: %s\n", mangled_name.c_str());
+      llvm_unreachable("failed to demangle name");
+#endif
+      return 0;
+    }
+    break;
+
+    // element types
+  case 'l':
+    blen <<= 1; // long
+  case 'i':
+    blen <<= 1; // int
+  case 's':
+    blen <<= 1; // short
+  case 'c':     // char
+  case 'a':     // signed char
+    type_info->type_id = Type::IntegerTyID;
+    type_info->is_signed = true;
+    type_info->byte_len = blen;
+    break;
+  case 'm':
+    blen <<= 1; // unsigned long
+  case 'j':
+    blen <<= 1; // unsigned int
+  case 't':
+    blen <<= 1; // unsigned short
+  case 'h':     // unsigned char
+    type_info->type_id = Type::IntegerTyID;
+    type_info->is_signed = false;
+    type_info->byte_len = blen;
+    break;
+  case 'd':
+    blen = 2; // double float
+  case 'f':
+    blen <<= 2; // single float
+    type_info->type_id = Type::FloatTyID;
+    type_info->is_signed = true;
+    type_info->byte_len = blen;
+    break;
+  case 'v': // void
+    break;
+  case '1':
+  case '2':
+  case '3':
+  case '4':
+  case '5':
+  case '6':
+  case '7':
+  case '8':
+  case '9': {
+    type_info->type_id = Type::StructTyID;
+    pos--;
+    type_info->name = GetUnmangledName(mangled_name, &pos);
+    break;
+  }
+  case '.':
+    return 0;
+  default:
+#ifdef DEBUG
+    printf("Func: %s\n", mangled_name.c_str());
+    llvm_unreachable("failed to demangle name");
+#endif
+    return 0;
+  }
+  return pos;
+}
+} // namespace
+
+////////////////////////////////////////////////////////////////////////////////
+// FunctionInfo::GetFromMangledNameCheck
+//   - parse mangled name as far as possible. Some names are an aggregate of
+//   fields separated by '.'
+//   - extract name and parameter types, and return type for 'convert' functions
+//   - return true if the mangled name can be fully parsed
+bool Builtins::FunctionInfo::GetFromMangledNameCheck(
+    const std::string &mangled_name) {
+  size_t pos = 0;
+  if (mangled_name[pos++] != '_' || mangled_name[pos++] != 'Z') {
+    name_ = mangled_name;
+    return false;
+  }
+
+  name_ = GetUnmangledName(mangled_name, &pos);
+
+  if (!name_.compare(0, 8, "convert_")) {
+    // deduce return type from name, only for convert
+    char tok = name_[8];
+    return_type_.is_signed = tok != 'u'; // unsigned
+    return_type_.type_id = tok == 'f' ? Type::FloatTyID : Type::IntegerTyID;
+  }
+
+  ParamTypeInfo prev_type_info;
+
+  auto mangled_name_len = mangled_name.length();
+  while (pos < mangled_name_len) {
+    ParamTypeInfo type_info;
+    if (mangled_name[pos] == 'S') {
+      // handle duplicate param_type
+      if (mangled_name[pos + 1] != '_') {
+        return false;
+      }
+      params_.push_back(prev_type_info);
```

Comment:
What if this is an invalid back-reference?  E.g. this S is for the first param.

In fact, could avoid the need for prev_type_info by indexing the last current value of params.
And then you also end up catching the malicious case, where params_ is empty.